rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,395 @@
+
+## Alternate shard structure
+
+Context: The structure is `pkgs/by-name/${shard}/${name}` with `shard` being the lowercased two-letter prefix of `name`.
+
+Alternatives:
+- A flat directory, where `pkgs.hello` would be in `pkgs/by-name/hello`.
+  - (+) Simpler for the user and code.
+  - (-) The GitHub web interface only renders the first 1 000 entries when browsing directories, which would make most packages inaccessible in this way.
+    - (+) This feature is not used often.
+      - (-) [A poll](https://discourse.nixos.org/t/poll-how-often-do-you-navigate-through-directories-on-github-to-get-to-a-nixpkgs-package-file/21641) showed that about 41% of people rely on this feature every week.
+  - (-) Bad because it makes `git` and file listings slower.
+- Use three-letter or four-letter prefixes.
+  - (-) Also leads to directories containing more than 1 000 entries, see above.
+- Use multi-level structure, e.g. a two-level two-letter prefix structure where `hello` is in `pkgs/by-name/he/ll/hello`
+  - (+) This would allow virtually a unlimited number of packages without performance problems
+  - (-) It's hard to understand, type and implement, needs a special case for packages with few characters
+    - E.g. `x` could go in `pkgs/by-name/x-/--/x`
+  - (-) There's not enough packages even in Nixpkgs that a two-level 4-letter structure would make sense. Most of the structure would only be filled by a couple entries.
+  - (-) Even Git only uses 2-letter prefixes for its objects hex hashes
+- Use two-letter prefixes split into two directories, like `pkgs/by-name/h/e/hello`
+  - (+) Allows easy traversal by clicking on GitHub file listings, shard directories being limited to under 40 children
+  - (-) Requires special-casing single-letter attribute names
+    - (+) There's currently only 6 such cases, which could be handled on a one-off basis
+  - (-) Makes auto-completion worse, having to tab-complete once more
+  - (-) Makes it harder to create shards: if a shard doesn't exist yet, it has to be created with either one or two `mkdir`'s, or a `mkdir -p`
+- Use a dynamic structure where directories are rebalanced when they have too many entries.
+  E.g. `pkgs.foobar` could be in `pkgs/by-name/f/foobar` initially.
+  But when there's more than 1 000 packages starting with `f`, all packages starting with `f` are distributed under 2-letter prefixes, moving `foobar` to `pkgs/by-name/fo/foobar`.
+  - (-) The structure depends not only on the name of the package then, making it harder to find packages again and figure out where they should go
+  - (-) Complex to implement
+
+## Alternate `package.nix` filename
+
+Context: The only file that has to exist in package directories is `package.nix`, it must contain a function suitable for `callPackage`.
+
+Alternatives:
+- `default.nix`
+  - (+) `default.nix` is already a convention most people are used to.
+  - (-) We don't benefit from the usual `default.nix` benefits:
```

Comment:
We would benefit that people are used to placing default.nix files in directories. It is the defacto filename which IMO should have just continued to be used.